src/Distribution/Server/Features/Vouch.hs

```diff
@@ -0,0 +1,218 @@
+          { restoreEntry = error "Unexpected backup entry"
+          , restoreFinalize = return initialVouchData
+          }
+  pure StateComponent
+    { stateDesc = "Keeps track of vouches"
+    , stateHandle = st
+    , getState = query st GetVouchesData
+    , putState = update st . ReplaceVouchesData
+    , backupState = \_ _ -> []
+    , restoreState = restore
+    , resetState = vouchStateComponent
+    }
+
+data VouchFeature =
+  VouchFeature
+    { vouchFeatureInterface :: HackageFeature
+    }
+
+instance IsHackageFeature VouchFeature where
+  getFeatureInterface = vouchFeatureInterface
+
+requiredCountOfVouches :: Int
+requiredCountOfVouches = 3
+
+isWithinLastMonth :: UTCTime -> (UserId, UTCTime) -> Bool
+isWithinLastMonth now (_, vouchTime) =
+  addUTCTime (30 * nominalDay) vouchTime < now
+
+data Err
+  = NotAnUploader
+  | You'reTooNew
+  | VoucheeAlreadyUploader
+  | AlreadySufficientlyVouched
+  | YouAlreadyVouched
+
+data Success = AddVouchComplete | AddVouchIncomplete
+
+judge :: Group.UserIdSet -> UTCTime -> UserId -> [(UserId, UTCTime)] -> [(UserId, UTCTime)] -> UserId -> Either Err (Either Err Success)
+judge ugroup now vouchee vouchersForVoucher existingVouchers voucher = runExceptT $ do
+  when (not (voucher `Group.member` ugroup)) $
+    throwError NotAnUploader
+  -- You can only vouch for non-uploaders, so if this list has items, the user is uploader because of these vouches.
+  -- Make sure none of them are too recent.
+  when (length vouchersForVoucher >= requiredCountOfVouches && any (isWithinLastMonth now) vouchersForVoucher) $
+    throwError You'reTooNew
+  when (vouchee `Group.member` ugroup) $
+    throwError VoucheeAlreadyUploader
+  when (length existingVouchers >= 3) $
+    throwError AlreadySufficientlyVouched
+  when (voucher `elem` map fst existingVouchers) $
+    throwError YouAlreadyVouched
+  pure $
+    if length existingVouchers == requiredCountOfVouches - 1
+       then AddVouchComplete
+       else AddVouchIncomplete
+
+renderToLBS :: (UserId -> ServerPartE UserInfo) -> [(UserId, UTCTime)] -> ServerPartE TemplateAttr
+renderToLBS lookupUserInfo vouches = do
+  rendered <- traverse renderVouchers vouches
+  pure $
+    templateUnescaped "vouches" $
+      if null rendered
+         then LBS.pack "Nobody has vouched yet."
+         else LBS.intercalate mempty rendered
+  where
+  renderVouchers :: (UserId, UTCTime) -> ServerPartE LBS.ByteString
+  renderVouchers (uid, timestamp) = do
+    info <- lookupUserInfo uid
+    let UserName name = userName info
+        -- We don't need to show millisecond precision
+        -- So we truncate it off here
+        truncated = truncate $ utctDayTime timestamp
+        newUTCTime = timestamp {utctDayTime = secondsToDiffTime truncated}
+    pure . toUTF8LBS . prettyHtmlFragment . li . stringToHtml $ name <> " vouched on " <> formatShow iso8601Format newUTCTime
+
+initVouchFeature :: ServerEnv -> IO (UserFeature -> UploadFeature -> IO VouchFeature)
```

Comment:
Is this callback setup required? can't you immediatly chug in UserFeatue and UploadFeature?

eg 
```
initVouchFeature :: ServerEnv -> UserFeature -> UploadFeature -> IO VouchFeature
```

Reply:
This is the pattern used across the codebase, look in `Server/Features.hs` (the use site) so see how it's used all over. I don't object to your observation but I think we should change it consistently in a different PR if it needs to change.

Reply:
it's fine, it's a minor style nitpick :)